### ISSUE TITLE
Starting from 13.0.0 browser address bar doesn't reflect a route pushed into a StatefulShellRoute branch while optionURLReflectsImperativeAPIs is set to true

### ISSUE THREAD

**Issue by alexlvn** [flutter/flutter#142053]
### What package does this bug report belong to?

go_router

### What target platforms are you seeing this bug on?

Web

### Have you already upgraded your packages?

Yes

### Dependency versions

<details><summary>pubspec.lock</summary>

```lock
# Generated by pub
# See https://dart.dev/tools/pub/glossary#lockfile
packages:
  async:
    dependency: transitive
    description:
      name: async
      sha256: "947bfcf187f74dbc5e146c9eb9c0f10c9f8b30743e341481c1e2ed3ecc18c20c"
      url: "https://pub.dev"
    source: hosted
    version: "2.11.0"
  boolean_selector:
    dependency: transitive
    description:
      name: boolean_selector
      sha256: "6cfb5af12253eaf2b368f07bacc5a80d1301a071c73360d746b7f2e32d762c66"
      url: "https://pub.dev"
    source: hosted
    version: "2.1.1"
  characters:
    dependency: transitive
    description:
      name: characters
      sha256: "04a925763edad70e8443c99234dc3328f442e811f1d8fd1a72f1c8ad0f69a605"
      url: "https://pub.dev"
    source: hosted
    version: "1.3.0"
  clock:
    dependency: transitive
    description:
      name: clock
      sha256: cb6d7f03e1de671e34607e909a7213e31d7752be4fb66a86d29fe1eb14bfb5cf
      url: "https://pub.dev"
    source: hosted
    version: "1.1.1"
  collection:
    dependency: transitive
    description:
      name: collection
      sha256: ee67cb0715911d28db6bf4af1026078bd6f0128b07a5f66fb2ed94ec6783c09a
      url: "https://pub.dev"
    source: hosted
    version: "1.18.0"
  fake_async:
    dependency: transitive
    description:
      name: fake_async
      sha256: "511392330127add0b769b75a987850d136345d9227c6b94c96a04cf4a391bf78"
      url: "https://pub.dev"
    source: hosted
    version: "1.3.1"
  flutter:
    dependency: "direct main"
    description: flutter
    source: sdk
    version: "0.0.0"
  flutter_lints:
    dependency: "direct dev"
    description:
      name: flutter_lints
      sha256: a25a15ebbdfc33ab1cd26c63a6ee519df92338a9c10f122adda92938253bef04
      url: "https://pub.dev"
    source: hosted
    version: "2.0.3"
  flutter_test:
    dependency: "direct dev"
    description: flutter
    source: sdk
    version: "0.0.0"
  flutter_web_plugins:
    dependency: "direct main"
    description: flutter
    source: sdk
    version: "0.0.0"
  go_router:
    dependency: "direct main"
    description:
      name: go_router
      sha256: "3b40e751eaaa855179b416974d59d29669e750d2e50fcdb2b37f1cb0ca8c803a"
      url: "https://pub.dev"
    source: hosted
    version: "13.0.1"
  lints:
    dependency: transitive
    description:
      name: lints
      sha256: "0a217c6c989d21039f1498c3ed9f3ed71b354e69873f13a8dfc3c9fe76f1b452"
      url: "https://pub.dev"
    source: hosted
    version: "2.1.1"
  logging:
    dependency: transitive
    description:
      name: logging
      sha256: "623a88c9594aa774443aa3eb2d41807a48486b5613e67599fb4c41c0ad47c340"
      url: "https://pub.dev"
    source: hosted
    version: "1.2.0"
  matcher:
    dependency: transitive
    description:
      name: matcher
      sha256: "1803e76e6653768d64ed8ff2e1e67bea3ad4b923eb5c56a295c3e634bad5960e"
      url: "https://pub.dev"
    source: hosted
    version: "0.12.16"
  material_color_utilities:
    dependency: transitive
    description:
      name: material_color_utilities
      sha256: "9528f2f296073ff54cb9fee677df673ace1218163c3bc7628093e7eed5203d41"
      url: "https://pub.dev"
    source: hosted
    version: "0.5.0"
  meta:
    dependency: transitive
    description:
      name: meta
      sha256: a6e590c838b18133bb482a2745ad77c5bb7715fb0451209e1a7567d416678b8e
      url: "https://pub.dev"
    source: hosted
    version: "1.10.0"
  path:
    dependency: transitive
    description:
      name: path
      sha256: "8829d8a55c13fc0e37127c29fedf290c102f4e40ae94ada574091fe0ff96c917"
      url: "https://pub.dev"
    source: hosted
    version: "1.8.3"
  sky_engine:
    dependency: transitive
    description: flutter
    source: sdk
    version: "0.0.99"
  source_span:
    dependency: transitive
    description:
      name: source_span
      sha256: "53e943d4206a5e30df338fd4c6e7a077e02254531b138a15aec3bd143c1a8b3c"
      url: "https://pub.dev"
    source: hosted
    version: "1.10.0"
  stack_trace:
    dependency: transitive
    description:
      name: stack_trace
      sha256: "73713990125a6d93122541237550ee3352a2d84baad52d375a4cad2eb9b7ce0b"
      url: "https://pub.dev"
    source: hosted
    version: "1.11.1"
  stream_channel:
    dependency: transitive
    description:
      name: stream_channel
      sha256: ba2aa5d8cc609d96bbb2899c28934f9e1af5cddbd60a827822ea467161eb54e7
      url: "https://pub.dev"
    source: hosted
    version: "2.1.2"
  string_scanner:
    dependency: transitive
    description:
      name: string_scanner
      sha256: "556692adab6cfa87322a115640c11f13cb77b3f076ddcc5d6ae3c20242bedcde"
      url: "https://pub.dev"
    source: hosted
    version: "1.2.0"
  term_glyph:
    dependency: transitive
    description:
      name: term_glyph
      sha256: a29248a84fbb7c79282b40b8c72a1209db169a2e0542bce341da992fe1bc7e84
      url: "https://pub.dev"
    source: hosted
    version: "1.2.1"
  test_api:
    dependency: transitive
    description:
      name: test_api
      sha256: "5c2f730018264d276c20e4f1503fd1308dfbbae39ec8ee63c5236311ac06954b"
      url: "https://pub.dev"
    source: hosted
    version: "0.6.1"
  vector_math:
    dependency: transitive
    description:
      name: vector_math
      sha256: "80b3257d1492ce4d091729e3a67a60407d227c27241d6927be0130c98e741803"
      url: "https://pub.dev"
    source: hosted
    version: "2.1.4"
  web:
    dependency: transitive
    description:
      name: web
      sha256: afe077240a270dcfd2aafe77602b4113645af95d0ad31128cc02bce5ac5d5152
      url: "https://pub.dev"
    source: hosted
    version: "0.3.0"
sdks:
  dart: ">=3.2.3 <4.0.0"
  flutter: ">=3.10.0"
```

</details>

### Steps to reproduce

1. Run the code sample using `go_router: ^13.0.1`
2. Click on the "Push /c" button

### Expected results

The path should be `/c`

### Actual results

The path is `/`

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_web_plugins/flutter_web_plugins.dart';
import 'package:go_router/go_router.dart';

void main() {
  usePathUrlStrategy();
  GoRouter.optionURLReflectsImperativeAPIs = true;
  runApp(const MyApp());
}

final goRouter = GoRouter(
  routes: [
    StatefulShellRoute.indexedStack(
      builder: (context, state, navigationShell) => _buildStatefulShellScaffold(navigationShell),
      branches: [
        StatefulShellBranch(
          routes: [
            GoRoute(
              path: '/',
              builder: (context, state) => const APage(),
              routes: [
                GoRoute(
                  path: 'c',
                  builder: (context, state) => const CPage(),
                ),
              ],
            ),
          ],
        ),
        StatefulShellBranch(
          routes: [
            GoRoute(
              path: '/b',
              builder: (context, state) => const BPage(),
            ),
          ],
        ),
      ],
    ),
  ],
);

Widget _buildStatefulShellScaffold(StatefulNavigationShell navigationShell) {
  return Scaffold(
    bottomNavigationBar: BottomNavigationBar(
      onTap: (i) => navigationShell.goBranch(i),
      currentIndex: navigationShell.currentIndex,
      items: const [
        BottomNavigationBarItem(
          icon: Icon(Icons.cake),
          label: 'A',
        ),
        BottomNavigationBarItem(
          icon: Icon(Icons.cookie),
          label: 'B',
        ),
      ],
    ),
    body: navigationShell,
  );
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      routerConfig: goRouter,
    );
  }
}

class APage extends StatelessWidget {
  const APage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('A')),
      body: Center(
        child: FilledButton(
          onPressed: () => context.push('/c'),
          child: const Text('Push /c'),
        ),
      ),
    );
  }
}

class BPage extends StatelessWidget {
  const BPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('B')),
      body: Center(
        child: FilledButton(
          onPressed: () => context.push('/c'),
          child: const Text('Push /c'),
        ),
      ),
    );
  }
}


class CPage extends StatelessWidget {
  const CPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('C')),
      body: const Center(child: Text('This is screen C')),
    );
  }
}
```

</details>

### Screenshots or Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

Reproduced with `go_router: ^13.0.1`:
<img width="598" alt="image" src="https://github.com/flutter/flutter/assets/157372677/e828b4ea-3d22-4f7f-b126-3c4364bde2ae">

URL updates as expected with `go_router: ^12.1.3`:
<img width="601" alt="image" src="https://github.com/flutter/flutter/assets/157372677/f1702126-8434-4688-836f-d840b63afe1c">

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel stable, 3.16.5, on macOS 14.1 23B74 darwin-arm64, locale en-US)
    • Flutter version 3.16.5 on channel stable at /Users/wok/fvm/versions/3.16.5
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.19.0/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/wok/fvm/versions/3.16.5. Consider adding /Users/wok/fvm/versions/3.16.5/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (5 weeks ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/wok/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.85.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (3 available)
    • Alexandr Levin’s iPhone (mobile) • 00008101-000E14300AC1001E • ios            • iOS 17.2.1 21C66
    • macOS (desktop)                  • macos                     • darwin-arm64   • macOS 14.1 23B74 darwin-arm64
    • Chrome (web)                     • chrome                    • web-javascript • Google Chrome 120.0.6099.234

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by huycozy** [comment]
Hi @alexlvn 
This doesn't seem to relate to `optionURLReflectsImperativeAPIs` as I still see the issue with its default value (false). 

Taking a look deeper at your code, I found that the cause is from `context.push` on page A to page C. You can try using `GoRouter.of(context).go` and see if it works as expected. So could you try and confirm if this is appropriate to your case?

**Comment by alexlvn** [comment]
Hi @huycozy 

> This doesn't seem to relate to optionURLReflectsImperativeAPIs as I still see the issue with its default value (false).

The default behavior of go_router is to not to reflect route changes made by imperative APIs like `GoRouter.of(context).push` in the browser address bar. Setting `optionURLReflectsImperativeAPIs` to `true` should make the URL update when using imperative APIs. It worked like this previously with `go_router: ^12.1.3` but doesn't work in the latest go_router versions. You can try to run my code sample using `go_router: ^12.1.3` and observe the browser address bar updating as expected. (or see the attached screenshot for `go_router: ^12.1.3`)

> Taking a look deeper at your code, I found that the cause is from context.push on page A to page C. You can try using GoRouter.of(context).go and see if it works as expected. So could you try and confirm if this is appropriate to your case?

`GoRouter.of(context).go` is from declarative APIs and works as expected, this issue is about imperative APIs.

**Comment by naveenbharadwaj19** [comment]
Even I face the same issue. I'm not using shell route. Using `context.pushReplacementNamed` or `context.replaceNamed` changes the URL to "/", and the `Router.neglect` is also not working. The browser is still catching the history.

```dart
context.pushReplacementNamed("dashboard");

Router.neglect(
    context,
    () => context.goNamed("bookScan", queryParameters: {
          "name": patientName,
          "phoneNumber": phoneNumber
        }));
```

**Comment by anisovdev** [comment]
Need help
I create Pull Request that fixes this issue, but reviewers are ignore this pull request. Maybe you knows who should review go_router Pull requests? Tag their in linked PR https://github.com/flutter/packages/pull/6236

**Comment by anisovdev** [comment]
fixed in 13.2.2
@darshankawar @huycozy  @alexlvn

**Comment by huycozy** [comment]
Thanks for your contribution, @anisovdev. Closing the issue as fixed!
